packages/controllers/src/permissions/PermissionController.test.ts

```diff
@@ -0,0 +1,4014 @@
+              caveats: [
+                {
+                  type: CaveatTypes.noopCaveat,
+                  value: 'bar',
+                },
+              ] as any,
+            },
+          },
+        }),
+      ).toThrow(new Error('NoopCaveat value must be null'));
+    });
+
+    it('throws if the requested permission specifies disallowed caveats', () => {
+      const controller = getDefaultPermissionController();
+      const origin = 'metamask.io';
+
+      expect(() =>
+        controller.grantPermissions({
+          subject: { origin },
+          approvedPermissions: {
+            wallet_getSecretObject: {
+              caveats: [
+                {
+                  type: CaveatTypes.filterArrayResponse,
+                  value: ['bar'],
+                },
+              ] as any,
+            },
+          },
+        }),
+      ).toThrow(
+        new errors.ForbiddenCaveatError(
+          CaveatTypes.filterArrayResponse,
+          origin,
+          PermissionNames.wallet_getSecretObject,
+        ),
+      );
+    });
+
+    it('throws if the requested permission specifies caveats, and no caveats are allowed', () => {
+      const controller = getDefaultPermissionController();
+      const origin = 'metamask.io';
+
+      expect(() =>
+        controller.grantPermissions({
+          subject: { origin },
+          approvedPermissions: {
+            wallet_doubleNumber: {
+              caveats: [
+                {
+                  type: CaveatTypes.filterArrayResponse,
+                  value: ['bar'],
+                },
+              ] as any,
+            },
+          },
+        }),
+      ).toThrow(
+        new errors.ForbiddenCaveatError(
+          CaveatTypes.filterArrayResponse,
+          origin,
+          PermissionNames.wallet_doubleNumber,
+        ),
+      );
+    });
+
+    it('throws if the permission validator throws', () => {
+      const controller = getDefaultPermissionController();
+      const origin = 'metamask.io';
+
+      expect(() =>
+        controller.grantPermissions({
+          subject: { origin },
+          approvedPermissions: {
+            [PermissionNames.wallet_noop2]: {
+              caveats: [
+                { type: CaveatTypes.filterArrayResponse, value: ['foo'] },
+              ],
+            },
+          },
+        }),
+      ).toThrow(new Error('noop permission validation failed'));
+    });
+  });
+
+  describe('requestPermissions', () => {
+    it('requests a permission', async () => {
+      const options = getPermissionControllerOptions();
+      const { messenger } = options;
+      const origin = 'metamask.io';
+
+      const callActionSpy = jest
+        .spyOn(messenger, 'call')
+        .mockImplementationOnce((async (...args: any) => {
+          const [, { requestData }] = args;
+          return {
+            metadata: { ...requestData.metadata },
+            permissions: { ...requestData.permissions },
+          };
+        }) as any);
```

Comment:
This cast does not appear to be necessary:
```suggestion
        .mockImplementationOnce(async (...args: any) => {
          const [, { requestData }] = args;
          return {
            metadata: { ...requestData.metadata },
            permissions: { ...requestData.permissions },
          };
        });
```